add support for private ipv4 addresses in subnet mapping of ELB V2

KadenLNelson opened the pull request:
Please let me know if there's anything else I should update, this is my first PR to `terraform-provider-aws`. I've tested this feature by creating an `aws_lb` like the example below. I've verified that the subnet mapping in the aws console shows the private ipv4 address.

### steps to test;
```hcl
provider "aws" {
  region  = "us-east-1"
  profile = "default"
}

resource "aws_vpc" "test" {
  cidr_block = "10.0.0.0/20"

  tags = {
    Name = "test-vpc"
  }
}

resource "aws_subnet" "test" {
  vpc_id     = "${aws_vpc.test.id}"
  cidr_block = "10.0.1.0/24"

  tags = {
    Name = "test-subnet"
  }
}

resource "aws_lb" "test" {
  name               = "test-lb-tf"
  internal           = true
  load_balancer_type = "network"
  subnet_mapping {
    subnet_id            = aws_subnet.test.id
    private_ipv4_address = "10.0.1.15"
  }
}
```

<!--- See what makes a good Pull Request at : https://github.com/terraform-providers/terraform-provider-aws/blob/master/.github/CONTRIBUTING.md#pull-requests --->

<!--- Please keep this note for the community --->

### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for pull request followers and do not help prioritize the request

<!--- Thank you for keeping this note for the community --->

<!--- If your PR fully resolves and should automatically close the linked issue, use Closes. Otherwise, use Relates --->
fixes https://github.com/terraform-providers/terraform-provider-aws/issues/11403
fixes https://github.com/terraform-providers/terraform-provider-aws/issues/11044

Release note for [CHANGELOG](https://github.com/terraform-providers/terraform-provider-aws/blob/master/CHANGELOG.md):
<!--
If change is not user facing, just write "NONE" in the release-note block below.
-->

```release-note
add support for private ipv4 address in subnet mappings of resource aws_lb
```

Output from acceptance testing:

<!--
Replace TestAccXXX with a pattern that matches the tests affected by this PR.

For more information on the `-run` flag, see the `go test` documentation at https://tip.golang.org/cmd/go/#hdr-Testing_flags.
-->
```
$ make testacc TEST=./aws TESTARGS='-run=TestAccAWSLB_NLB_privateipv4address'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -count 1 -parallel 20 -run=TestAccAWSLB_NLB_privateipv4address -timeout 120m
=== RUN   TestAccAWSLB_NLB_privateipv4address
=== PAUSE TestAccAWSLB_NLB_privateipv4address
=== CONT  TestAccAWSLB_NLB_privateipv4address
--- PASS: TestAccAWSLB_NLB_privateipv4address (228.60s)
PASS
ok      github.com/terraform-providers/terraform-provider-aws/aws       228.635s
```

ewbankkit commented:
Closes https://github.com/terraform-providers/terraform-provider-aws/issues/11044.

KadenLNelson commented:
/assign KadenLNelson

cSNiHab commented:
Hi, we're currently deciding if we should wait for this PR to be merged or if we should implement a null_resource workaround to get private static IPs in.

Could you give us an indication if this PR will be merged (and if so, are there any timelines?)

KadenLNelson commented:
Currently awaiting maintainer review.

TimJDFletcher commented:
We have just run in this problem during a roll out of access to our on-premise resources from AWS.

Any news from the code review on this?

KadenLNelson commented:
> We have just run in this problem during a roll out of access to our on-premise resources from AWS.
> 
> Any news from the code review on this?

@bflad

florianlocqueneux commented:
Hello,
Any chance to get this reviewed ?
It could be really help full to get this merged.

ernstvonoelsen commented on website/docs/r/lb.html.markdown:
```diff
@@ -74,6 +74,25 @@ resource "aws_lb" "example" {
+
+  subnet_mapping {
+    subnet_id     = "${aws_subnet.example2.id}"
+    private_ivp4_address = "10.0.2.15"
```
typo: `private_ivp4_address` vs `private_ipv4_address`

ernstvonoelsen commented on website/docs/r/lb.html.markdown:
```diff
@@ -74,6 +74,25 @@ resource "aws_lb" "example" {
+
+  subnet_mapping {
+    subnet_id     = "${aws_subnet.example1.id}"
+    private_ivp4_address = "10.0.1.15"
```
typo: private_ivp4_address vs private_ipv4_address

florianlocqueneux replied:
Hello, Is there any chance to get this released ? Many thanks for your review and comments on this, would be really helpfully to see this feature implemented :)

bmathieu33 commented:
Should this PR  also mark https://github.com/terraform-providers/terraform-provider-aws/issues/3007 as closed?

ewbankkit commented:
@bmathieu33 It looks like #3007 is more related to #2901 / #11000 but given the decision on that solution, yes it should probably be closed (it does contain a lot of great workarounds but will still be searchable). Thanks.

MuhammadShaban commented:
is there any timelines for this feature?

florianlocqueneux commented:
This improvement would be really welcome here :)
Any visibility on the timeline would help as well.
Thank you so much

florianlocqueneux commented:
Hi  @KadenLNelson , unfortunately, your last commit is not successful.

dthvt commented:
The CI job is not particularly forthcoming w/ the root cause of the failure.  Is it because the `=` signs aren't aligned in the documentation HCL snippet?  That's about the only thing I can see that looks like it might be a lint failure.

KadenLNelson commented:
- I fixed the indentation in the documentation and added an acceptance test. 

- I accidentally removed the review request from `terraform-providers/aws-provider`, and I don't know how I can add it back.

@maryelizbeth when is this PR slated to be merged?

ericrichtert commented:
@maryelizbeth or @KadenLNelson 
Who can provide me a release date for this?

KadenLNelson commented:
> @maryelizbeth or @KadenLNelson 
> Who can provide me a release date for this?

I'm unaware of when this change may be merged in, a member of the Hashicorp team may be able to shed more light on this.

dexterlakin commented:
Any ideas as to when this will get merged? It's not a massive change.

I have a requirement to deploy an AWS NLB with a static private IP address at work so I'd like this resolved.

SureshGoli123 commented:
We also had urgent requriement to put static_ips for NLBs. And we are heavly dependent on the terraform. In which release this problem will get fixed?

Monpoke commented:
Hello ! 
We were looking to add static ips to our NLB using Terraform provider.
That's funny we just found this issue and last comment is one hour ago!

#livemylife

Thanks for the work @KadenLNelson!

smolar commented:
Maybe this feature is not available in all regions or something, parameter `PrivateIpv4Address` isn't in official documentation (https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-elasticloadbalancingv2-loadbalancer.html). 

I don't know, I really needed it and I had no more time to wait, so here is my workaround using `aws_cloudformation_stack`

```
resource "aws_cloudformation_stack" "example" {

  name          = var.nlbname
  template_body = file("./templates/nlb-cf.json")

  parameters = {
    Name                 = var.nlbname
    Scheme               = "internal"
    IpAddressType        = "ipv4"
    Subnet1a             = aws_subnet.example1a.id
    Subnet1b             = aws_subnet.example1b.id
    PrivateIPv4Address1a = 1.2.3.254
    PrivateIPv4Address1b = 1.2.4.254
    Type                 = "network"
    TagNameValue         = var.nlbname
  }
}

resource "aws_lb_listener" "ListenerExample" {
  load_balancer_arn = aws_cloudformation_stack.example.outputs["arnlb"]
  port              = "8000"
  protocol          = "TCP"

  default_action {
    type             = "forward"
    target_group_arn = aws_lb_target_group.example.arn
  }

  depends_on = [aws_cloudformation_stack.example]
}

./templates/nlb-cf.json:
{
  "Description": "NLB cloudformation template",
  "Parameters": {
    "Name": {
      "Type": "String",
      "Description": "NLB Name"
    },
    "Scheme": {
      "Type": "String",
      "Description": "Scheme Type"
    },
    "IpAddressType": {
      "Type": "String",
      "Description": "IP Address Type"
    },
    "Subnet1a": {
      "Type": "String",
      "Description": "Subnet ID"
    },
    "Subnet1b": {
      "Type": "String",
      "Description": "Subnet ID"
    },
    "PrivateIPv4Address1a": {
      "Type": "String",
      "Description": "PrivateIpv4Address cidr 1a"
    },
    "PrivateIPv4Address1b": {
      "Type": "String",
      "Description": "PrivateIpv4Address cidr 1b"
    },
    "Type": {
      "Type": "String",
      "Description": "LB Type"
    },
    "TagNameValue": {
      "Type": "String",
      "Description": "LB Tags"
    }
  },
  "Resources": {
    "NLB": {
      "Type" : "AWS::ElasticLoadBalancingV2::LoadBalancer",
      "Properties" : {
          "Name" : { "Ref": "Name" },
          "Scheme" : { "Ref": "Scheme" },
          "IpAddressType" : { "Ref": "IpAddressType" },
          "SubnetMappings" : [
          {
            "SubnetId": { "Ref": "Subnet1a" },
            "PrivateIPv4Address": { "Ref": "PrivateIPv4Address1a" }
          },
          {
            "SubnetId": { "Ref": "Subnet1b" },
            "PrivateIPv4Address": { "Ref": "PrivateIPv4Address1b" }
          }
          ],
          "Type" : { "Ref": "Type" },
          "Tags" : [
          {
            "Key" : "Name",
            "Value" : { "Ref": "TagNameValue" }
          }
        ]
      }
    }
  },
  "Outputs" : {
    "arnlb" : {
      "Description" : "ARN of NLB",
      "Value" : { "Ref": "NLB" }
    }
  }
}
```

frittentheke commented:
> Maybe this feature is not available in all regions or something, parameter `PrivateIpv4Address` isn't in official documentation (https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-elasticloadbalancingv2-loadbalancer.html).

Certainly I haven't check all regions, but according to the discussion on the previous PR back in 2018, there already was some wondering about that: https://github.com/terraform-providers/terraform-provider-aws/pull/2901#issuecomment-372851295

dthvt commented:
I suspect this is a case of CloudFormation lagging behind the service API.  ELBv2 NLB API includes support for it and as far as I know this is not region limited. https://docs.aws.amazon.com/elasticloadbalancing/latest/APIReference/API_SubnetMapping.html

smolar commented:
Ok, then i don't know why this feature is not released yet :)

kazz187 commented:
ping

We need this feature right now💨

RJPercival commented:
@KadenLNelson, can you re-request review from the current reviewers? They might just be overlooking its existence.

KadenLNelson commented:
Mary's response to my inquiry about the status of this pull request

> Hey Kaden, 
> 
> As you may have gathered, the AWS team is quite small. For this reason among others, we've tried to increase the transparency about what we're working on. In addition to items on our roadmap, we make space in each release for community pull requests and technical debt. These items are tracked in our team projects as well as HashiCorp's internal tracking system. 
> 
> Your pull request is on our project board and will be prioritized for review this quarter. We recognize the significant community support and conversation behind it, and once we are ready to review it one of our engineers will engage with you for review. Once the PR has been reviewed, we will add it to a release milestone indicating that it will ship. 
> 
> As for what you can do to ensure a smooth review process, review our contributing guide and ensure that the PR meets the expectations set in the contribution checklist. If you've already done this, great! We'll review the PR and leave any relevant comments so ensure you have notifications on to receive and respond to comments in a timely fashion. 
> 
> Hope this helps to clear up any questions! Let me know if there's anything that needs clarification. 
> 
> - M

DrFaust92 commented on aws/resource_aws_lb.go:
```diff
@@ -115,6 +115,11 @@ func resourceAwsLb() *schema.Resource {
+						"private_ipv4_address": {
+							Type:     schema.TypeString,
+							Optional: true,
+							ForceNew: true,
```
add validation here `ValidateFunc: validation.IsIPv4Address,`

DrFaust92 commented on aws/resource_aws_lb_test.go:
```diff
@@ -191,6 +191,40 @@ func TestAccAWSLB_networkLoadbalancerEIP(t *testing.T) {
+			{
+				Config: testAccAWSLBConfig_networkLoadBalancerPrivateIPV4Address(lbName),
+				Check: resource.ComposeAggregateTestCheckFunc(
+					testAccCheckAWSLBExists(resourceName, &conf),
```
most of these check are irrelevant to the specific case, lets leave only the relevant check
```
resource.TestCheckResourceAttr(resourceName, "internal", "true"),
resource.TestCheckResourceAttr(resourceName, "load_balancer_type", "network"),
resource.TestCheckResourceAttr(resourceName, "subnet_mapping.#", "1"),
```

DrFaust92 commented on aws/resource_aws_lb_test.go:
```diff
@@ -191,6 +191,40 @@ func TestAccAWSLB_networkLoadbalancerEIP(t *testing.T) {
+					resource.TestCheckResourceAttrSet(resourceName, "zone_id"),
+					resource.TestCheckResourceAttr(resourceName, "subnet_mapping.#", "1"),
+				),
+			},
```
add import step:
```
			{
				ResourceName:      resourceName,
				ImportState:       true,
				ImportStateVerify: true,
			},
```

DrFaust92 commented on aws/resource_aws_lb_test.go:
```diff
@@ -1770,6 +1804,46 @@ resource "aws_eip" "lb" {
+  vpc_id                  = "${aws_vpc.alb_test.id}"
+  cidr_block              = "10.10.0.0/21"
+  map_public_ip_on_launch = true
+  availability_zone       = "us-west-2a"
```
to make test region agnostic and to avoid environment specific failures, can please add the following to the test config:

```
data "aws_availability_zones" "available" {
  state = "available"

  filter {
    name   = "opt-in-status"
    values = ["opt-in-not-required"]
  }
}
```

and change the hardcoded az to `"${data.aws_availability_zones.available.names[0]}"`

DrFaust92 commented:
also Closes #11674

abohne commented:
@KadenLNelson Are you able to make the requested changes?

Floble commented:
Any progress on this? Is there an approximate time schedule indicating when to expect the merge? We would appreciate this solution.

bflad commented:
The maintainers are actively working on the 3.0 major release for the next week or two, then this is up for final review. If we do not see updates from @KadenLNelson to adjust the requested changes in the meantime, we will make the changes on top of this contribution to merge it in for release in 3.1.

TimJDFletcher commented:
We have just upgraded to the aws provider version 3.1 to try and use this functionality.

We are now getting the following errors at the plan stage:

`Error: error setting subnet_mapping: Invalid address to set: []string{"subnet_mapping", "0", "private_ipv4_address"}
`

I believe it could be related to this type of bug: https://github.com/terraform-providers/terraform-provider-aws/issues/4986

Should I open a new bug for this?

bflad commented:
@TimJDFletcher yes please create a new [bug report](https://github.com/terraform-providers/terraform-provider-aws/issues/new?template=Bug_Report.md). Thanks.
